### PR TITLE
Backblaze B2 Backend

### PR DESCRIPTION
**Pull request by powerpaul17**
I tried to implement a backend for Backblaze B2. I tested it a bit and it seems to run, but as I am not (yet) that familiar with python and this project, I am hoping for comments/corrections/suggestions.

### CODE REVIEW

File: src/s3ql/backends/b2/b2_backend.py [hunk 1]
```diff
@@ -0,0 +1,768 @@
...
+              ((500 <= exc.status <= 599) or
+               exc.status == 408 or
+               exc.status == 429)):
+            return True
+
+        elif (isinstance(exc, HTTPError) and
+              exc.status == 408):
+            return True
+
+        elif isinstance(exc, StateError):
```
Comment (powerpaul17) [inline src/s3ql/backends/b2/b2_backend.py]: I am not sure about this, I got a `StateError` occasionally and this seemed to fix it.

**Comment by powerpaul17** [comment]
Also, I would like to continue my work by writing more tests for the B2 backend, also a mock server to test the behavior against different server error responses.

**Comment by powerpaul17** [comment]
Also, is there a way to store additional data of the backend in the metadata database? Backblaze sometimes needs its internal file id for some operations and requesting the information takes an additional transaction, it would be very convenient if it could be stored somewhere after upload.

**Comment by Nikratio** [comment]
> Also, is there a way to store additional data of the backend in the metadata database? Backblaze sometimes needs its internal file id for some operations and requesting the information takes an additional transaction, it would be very convenient if it could be stored somewhere after upload.

Currently there is no way to do that. This functionality could be added, but the bar for that is very high, since we're adding extra complexity that is only used by a single backend, and the people who would benefit from this (users of the new backend) are potential future users, while the users that are paying the price already exist.

File: src/s3ql/backends/b2/b2_backend.py (hunk 1 above)
Comment (Nikratio) [reply src/s3ql/backends/b2/b2_backend.py]: If you get a `StateError`, this means there is a bug in how you're using dugong, or in dugong itself. In either case, the above is a workaround that hides the real problem. So no, this isn't the a good way to do it.

**Comment by Nikratio** [comment]
I had a glance over the code, and there's nothing that looks obviously bad/wrong. Unfortunately I won't have time for a full review in the near future.

File: src/s3ql/backends/b2/b2_backend.py (hunk 1 above)
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: That's what I was thinking. I had a short look into the dugong documentation (although due to lack of time not very extensively I admit), but it didn't completely clear up the way it works, so I'll dig into that a bit more. Do you have by chance some hints how to log/debug this?

**Comment by powerpaul17** [comment]
Thanks for your quick answers!

> Currently there is no way to do that. This functionality could be added, but the bar for that is very high, since we're adding extra complexity that is only used by a single backend, and the people who would benefit from this (users of the new backend) are potential future users, while the users that are paying the price already exist.

Understandable, another way would be caching the file ids locally but I guess that could lead to memory problems, so I leave it the way it is for now..

> I had a glance over the code, and there's nothing that looks obviously bad/wrong. Unfortunately I won't have time for a full review in the near future.

I understand, I also have work to do, I'll test in the meantime, and maybe improve some bits. And I hope we will get this merged sometime.

**Comment by szepeviktor** [comment]
@powerpaul17 Graz ist in der Nähe.

**Comment by szepeviktor** [comment]
@powerpaul17 Thank you for your work.
Do you plan to add a backend test? https://github.com/s3ql/s3ql/blob/master/tests/t1_backends.py#L23-L26

**Comment by szepeviktor** [comment]
@powerpaul17 This container tests your work
https://github.com/szepeviktor/debian-server-tools/blob/master/virtualization/python-s3ql-test-b2/Dockerfile#L46

There is one failing test in it.

> ResourceWarning: unclosed <ssl.SSLSocket fd=13, family=AddressFamily.AF_INET, type=2049, proto=6, laddr=('172.17.0.2', 47322), raddr=('45.11.36.79', 443)>

```
================================================================================== FAILURES ===================================================================================
__________________________________________________________________________ test_copy[remote-b2/raw] ___________________________________________________________________________
Traceback (most recent call last):
  File "/s3ql/tests/pytest_checklogs.py", line 140, in pytest_runtest_call
    check_output(item)
  File "/s3ql/tests/pytest_checklogs.py", line 132, in check_output
    check_test_output(capmethod, item)
  File "/s3ql/tests/pytest_checklogs.py", line 106, in check_test_output
    raise AssertionError('Suspicious output to stderr (matched "%s")' % hit.group(0))
AssertionError: Suspicious output to stderr (matched "Warning")
---------------------------------------------------------------------------- Captured stderr call -----------------------------------------------------------------------------
/s3ql/src/s3ql/backends/common.py:143: ResourceWarning: unclosed <ssl.SSLSocket fd=13, family=AddressFamily.AF_INET, type=2049, proto=6, laddr=('172.17.0.2', 47322), raddr=('45.11.36.79', 443)>
  interval = exc.retry_after
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
================================================================== 1 failed, 78 passed, 4 skipped in 35.76s ===================================================================
```

**Comment by powerpaul17** [comment]
> @powerpaul17 Thank you for your work.
> Do you plan to add a backend test? https://github.com/s3ql/s3ql/blob/master/tests/t1_backends.py#L23-L26

I initially was planning to do it but it requires a separate b2 mocking server which I had no time to do yet.

The failing test is failing some times, but sometimes not (and it is not always the same one which fails..), I can successfully run all the tests separately. Unfortunately I have no clue about the warning/error but the backend is working fine on my server so far.

I rebased the branch onto the current master but it seems not to run due to a deprecation warning of the trio package, until this is sorted out I would recommend you try the b2_old branch (which I am using on my server at t he moment)

**Comment by szepeviktor** [comment]
@Nikratio Could you help us in the "unclosed ssl.SSLSocket" error?

**Comment by Nikratio** [comment]
This means that the backblaze backend is initializing an ssl.SSLSocket object somewhere, but then never closes it explicitly (probably instead relying on implicit closing when the object goes out of scope).

**Comment by powerpaul17** [comment]
You were right, I forgot to close the connection when invalidating a new upload url.

**Comment by szepeviktor** [comment]
All green 🍏 

kép

**Comment by szepeviktor** [comment]
@Nikratio Is it okay to merge?

**Comment by Nikratio** [comment]
What's the status of the issue from https://github.com/s3ql/s3ql/pull/116#issuecomment-517032565?

**Comment by powerpaul17** [comment]
> What's the status of the issue from [#116 (comment)](https://github.com/s3ql/s3ql/pull/116#issuecomment-517032565)?

Since it's not possible to save the internal file ids of Backblaze an additional HEAD request is sent to get the id.

File: src/s3ql/backends/b2/b2_error.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+            # Force 1s waiting time before retry
+            self.retry_after = 1
+
+    def __str__(self):
+        return '%s : %s - %s' % (self.status, self.code, self.message)
+
+
+class BadDigestError(B2Error): pass
+
+def _parse_retry_after_header(header):
```
Comment (Nikratio) [inline src/s3ql/backends/b2/b2_error.py]: Could this be de-duplicated against the similar function in backends/s3c.py

File: src/s3ql/backends/b2/b2_backend.py [hunk 3]
```diff
@@ -0,0 +1,807 @@
+import base64
```
Comment (Nikratio) [inline src/s3ql/backends/b2/b2_backend.py]: Please add a file docstring with copyright information

**Comment by Nikratio** [comment]
I do not have a lot of time for working on S3QL anymore. However, I do not think that my availability should limit work on S3QL.

If you can find another developer who has previously committed a significant amount of code to S3QL to review this and give his stamp of approval (and have the unit tests pass, of course), then I will merge this.

Glancing over https://github.com/s3ql/s3ql/graphs/contributors, d--j@ would probably be the best choice (if he is willing).

**Comment by Nikratio** [comment]
...one more condition, actually: someone needs to commit to supporting this backend in the future (no "fire and forget" features).

**Comment by powerpaul17** [comment]
> ...one more condition, actually: someone needs to commit to supporting this backend in the future (no "fire and forget" features).

Since I use it already on a daily basis, there is also an incentive for me to continue working on it. Although I cannot speak for eternity.. :wink:

**Comment by powerpaul17** [comment]
@d--j , @Nikratio suggested you could take a look at this PR as he is limited in time. Would that be possible for you? :slightly_smiling_face:

File: src/s3ql/backends/b2/b2_backend.py [hunk 4]
```diff
@@ -0,0 +1,807 @@
...
+            value = metadata[key]
+            if (not isinstance(value, (str, bytes, int, float, complex, bool))
+                and value is not None):
+                raise ValueError('value for key %s (%s) is not elementary' % (key, value))
+
+            if isinstance(value, (bytes, bytearray)):
+                value = base64.b64encode(value)
+
+            buffer = ('%s: %s' % (repr(key), repr(value)))
+            buffer = urllib.parse.quote(buffer, safe='!@#$^*()=+/?-_\'"><\\| `.,;:~')
```
Comment (d--j) [inline src/s3ql/backends/b2/b2_backend.py]: Why is this necessary? `_add_b2_metadata_to_headers` also calls `_b2_url_encode` on the value, doesn't it?
```suggestion
```

File: src/s3ql/backends/b2/b2_backend.py [hunk 5]
```diff
@@ -0,0 +1,807 @@
...
+                i = 0
+                while i * chunksize < len(buffer):
+                    k = 'meta-%03d' % info_header_count
+                    v = buffer[i * chunksize : (i + 1) * chunksize]
+                    metadata_dict[k] = v
+                    i += 1
+                    info_header_count += 1
+
+        # Backblaze B2 only allows 10 metadata headers (2 are used by format and md5)
+        assert info_header_count <= 8
```
Comment (d--j) [inline src/s3ql/backends/b2/b2_backend.py]: Since Backblaze can only handle 8 custom headers, we should probably be more efficient with them. This method will start a new header for every new metadata key. So we can only have at max 8 metadata keys. Works for now but it can easily break later down the line.

I think you should refactor this so that the metadata gets chunked into 2048 bytes chunks after you assembled the whole `repr`-buffer for the whole dict.

File: src/s3ql/backends/b2/b2_backend.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+            raise CorruptedObjectError('invalid metadata format: %s' % format_)
+
+        parts = []
+        for i in count():
+            part = headers.get('%smeta-%03d' % (self.info_header_prefix, i), None)
+            if part is None:
+                break
+            parts.append(part)
+
+        buffer = urllib.parse.unquote(','.join(parts))
```
Comment (d--j) [inline src/s3ql/backends/b2/b2_backend.py]:
```suggestion
        buffer = urllib.parse.unquote(''.join(parts))
```

When you currently safe more than 2048 bytes into a metadata key, this will not work. It will stick a `,` into a place where it does not belong. See https://github.com/s3ql/s3ql/blob/d00446705963317a6db5706228de110411918f49/src/s3ql/backends/s3c.py#L371 - there S3QL adds the `,` directly in the metadata. You need to do that, too.

This of course breaks all your currently persisted metadata. One could add a special case here for your legacy data (try to use `''.join(parts)` first, then `','.join(parts)` when that fails) but I wouldn't want to see those legacy codes in S3QL and I think @Nikratio wouldn't either. 
I think it's better that you re-write all metadata of your existing S3QL filesystems with a little upgrade-script.

**Comment by Nikratio** [comment]
> @Nikratio: @powerpaul17 has added the B2 backend in multiple files in a `backends/b2` subfolder. All other backends are in a single python file. I actually do not mind and would vote for refactoring all the other backends that way, too (eventually). But what do you think?

Daniel, thanks a lot for reviewing this!

I'm fine adding this in multiple files. I don't think we should refactor the existing code unless we need to do so for other reasons. I think either solution is fine, so turning one into the other is mostly churn that makes `git blame` less useful.

**Comment by powerpaul17** [comment]
@d--j, thank you very much for your review and the detailed comments. I'll try to address those as soon as possible but unfortunately holidays are over for me as well.. :wink:

File: src/s3ql/backends/b2/b2_backend.py (hunk 5 above)
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: I tried this back then but I did not succeed at that time (probably because I do not have a lot of python experience), the problem now is that I'm already using the backend with the metadata in this way since almost half a year.. and changing would need some conversion for my data..

File: src/s3ql/backends/b2/b2_backend.py (hunk 6 above)
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: > I think it's better that you re-write all metadata of your existing S3QL filesystems with a little upgrade-script.

Okay, that was what I was afraid of.. :wink:

File: src/s3ql/backends/b2/b2_backend.py [hunk 7]
```diff
@@ -0,0 +1,807 @@
...
+    def _reset_connections(self):
+        if self.api_connection:
+            self.api_connection.reset()
+
+        if self.download_connection:
+            self.download_connection.reset()
+
+    def _get_download_connection(self):
+        if self.download_connection is None:
+            self._authorize_account()
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: The authorization request returns `downloadUrl` and `apiUrl` which are then used to create the connection. So, if there is no connection it gets created by calling `self._authorize_account()`. But I suppose I can extract this into separate functions to avoid calling the API again. But this case should actually never happen because both connection should be created and used until the authorization token expires.

File: src/s3ql/backends/b2/b2_backend.py [hunk 8]
```diff
@@ -0,0 +1,807 @@
...
+
+        return self.download_connection
+
+    def _get_api_connection(self):
+        if self.api_connection is None:
+            self._authorize_account()
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: see comment above

File: src/s3ql/backends/b2/b2_backend.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (d--j) [reply src/s3ql/backends/b2/b2_backend.py]: You are calling `_close_connections` in `open_read`'s error case. That would benefit from not re-authorising.

File: rst/backends.rst [hunk 9]
```diff
@@ -374,6 +374,67 @@ The S3 compatible backend accepts the following backend options:
...
+Backblaze B2 is a cloud storage with its own API.
+
+The storage URL for backblaze b2 storage is ::
+
+   b2://<bucket-name>[/<prefix>]
+
+*bucket-name* is an (existing) bucket which has to be accessible with
+the provided account key. The *prefix* will be appended to all names
+used by S3QL and can be used to hold separate S3QL repositories in the
+same bucket.
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply rst/backends.rst]: Yes, I use Application Keys for my buckets, which are restricted to one bucket. I'll add more documentation here.

File: src/s3ql/backends/b2/b2_backend.py [hunk 10]
```diff
@@ -0,0 +1,807 @@
...
+import re
+import ssl
+import binascii
+from ast import literal_eval
+from itertools import count
+import io
+import os
+from shutil import copyfileobj
+
+from  typing import Dict, Tuple, Union, List
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: Is it possible to start using PEP 484, since the minimum python version is 3.6 now?

File: rst/backends.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (powerpaul17) [reply rst/backends.rst]: > or a key restricted to one bucket but then _authorize_account also needs to fetch the bucketID from the authentication response

So, do you mean to fetch the bucket's name entirely through authorization with the application key and not having to include it in the storage URL in this case?

File: src/s3ql/backends/b2/b2_backend.py [hunk 11]
```diff
@@ -0,0 +1,807 @@
...
+    @retry
+    @copy_ancestor_docstring
+    def copy(self, src, dest, metadata=None):
+        log.debug('started with %s, %s', src, dest)
+
+        source_file_id, source_file_name = self._get_file_id_and_name(src)
+
+        request_dict = {
+            'sourceFileId': source_file_id,
+            'fileName': self.prefix + dest.replace('\\', '=5C') # self._get_key_with_prefix(dest)
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: Because `_get_key_with_prefix` uses `_b2_url_encode` which in turn calls `urllib.parse.quote`, but in this case, the parameters are passed as JSON to B2, where only the backslashes have to be converted but not the rest.

File: rst/backends.rst (hunk 9 above)
[the remark this answers is not in the document]
Comment (d--j) [reply rst/backends.rst]: > So, do you mean to fetch the bucket's name entirely through authorization with the application key and not having to include it in the storage URL in this case?

No, but as far as I understood the B2 API docs you are not allowed to do `listBuckets` when you have a bucket-restricted application key. But even if you can call `b2_list_buckets`, when you use restricted application keys the call to `b2_authorize_account` will give you `allowed.bucketId` and `allowed.bucketName`. So no need to call `b2_list_buckets` to find out the bucket id in that case (one less API call).

File: src/s3ql/backends/b2/b2_backend.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (d--j) [reply src/s3ql/backends/b2/b2_backend.py]: > Is it possible to start using PEP 484, since the minimum python version is 3.6 now?

@Nikratio what do you think about using [PEP 484](https://www.python.org/dev/peps/pep-0484/)? Incremental (new/changed code only)? Refactor everything in one go? Not using type hints at all?

File: rst/backends.rst [hunk 12]
```diff
@@ -374,6 +374,67 @@ The S3 compatible backend accepts the following backend options:
...
+
+.. option:: account-id
+
+   Some API calls need the account id which can be passed by using
+   this parameter. If it is not provided, a separate call has to be
+   made to fetch it from the backend.
```
Comment (d--j) [inline rst/backends.rst]: Just realised, that this option is not needed (anymore?). The `accountId` gets returned from `b2_authorize_account` so no need for an extra API call.

Comment (powerpaul17) [reply rst/backends.rst]: You are right. :slightly_smiling_face:

**Comment by szepeviktor** [comment]
Very nice progress.

File: src/s3ql/backends/b2/b2_error.py [hunk 13]
```diff
@@ -19,14 +18,7 @@ def __init__(self, status, code, message, headers=None):
             self.retry_after = 1
```
Comment (d--j) [inline src/s3ql/backends/b2/b2_error.py]: no need for `_parse_retry_after_header` here, just change it to
```python
        # Force 1s waiting time before retry
        if not self.retry_after:
            self.retry_after = 1
```
since `super().__init__()` did that already.

File: src/s3ql/backends/b2/b2_backend.py (hunk 5 above)
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: I just read that while there is the 10 info header limitation, Backblaze imposes no limit on the size of the headers. That would render the `chunksize` essentially unnecessary in this case. But the overall size of the headers can not exceed 7000 bytes. Though I don't know exactly what is counted towards that limit and/or we can count that.

File: src/s3ql/backends/b2/b2_backend.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Nikratio) [reply src/s3ql/backends/b2/b2_backend.py]: I think adding type hints would be great.

File: src/s3ql/backends/b2/b2_backend.py (hunk 5 above)
Comment (d--j) [reply src/s3ql/backends/b2/b2_backend.py]: > I just read that while there is the 10 info header limitation, Backblaze imposes no limit on the size of the headers. That would render the `chunksize` essentially unnecessary in this case. [...]

It's probably safer to do chunking nevertheless – there could be crappy proxy servers in between S3QL and the B2 backend that do not like long header lines. But we could skip chunking for now and maybe introduce it when there is a user that needs chunking because of her outdated proxy infrastructure. (so add only one `X-Bz-Info-Meta-001` header – and when we decide to add chunking afterwards, that can be added in a backwards compatible manner).

> [...] But the overall size of the headers can not exceed 7000 bytes. [...]

I think 7000 bytes is quite enough. Here on my Swift Containers S3QL uses ~500 bytes of header data. Data objects store six metadata keys already (`encryption`, `signature`, `data`, `object_id`, `nonce` and `compression`) that and the two custom headers `Meta-Md5` and `Meta-Format` are eight custom headers that S3QL is using right now (of the 10 that are allowed).

> [...] Though I don't know exactly what is counted towards that limit and/or how we can count that.

from https://www.backblaze.com/b2/docs/files.html#httpHeaderSizeLimit
> HTTP Header Size Limit
The file name and file info must fit, along with the other necessary headers, within an 8KB limit imposed by some web servers and proxies. To ensure this, both now and in the future, B2 limits the combined header size for the file name and all file info to 7000 bytes. This limit applies to the fully encoded HTTP header line, including the carriage-return and newline. This header line is counted as 40 bytes:
> 
>    X-Bz-File-Name: %E8%87%AA%E7%94%B1.txt\r\n

(I am counting 41 bytes there, btw)

So `X-Bz-File-Name` and all `X-Bz-Info-Meta-*` headers count towards this 7000-bytes limit. S3QL might take something like 500 bytes of this limit for now. Leaves 6500 bytes for the file name – S3QL uses files names < 50 bytes so even with a very very long prefix we have a healthy safety margin.

Since it's not trivial to calculate the header size in advance and the limit is far bigger than what we need for now, I would skip adding an assertion for the 7000-byte limit. The B2 API will probably return a descriptive HTTP-Error anyways in this case.

File: src/s3ql/backends/b2/b2_backend.py [hunk 14]
```diff
@@ -0,0 +1,807 @@
...
+
+            raise
+
+        except (ConnectionClosed, ConnectionTimedOut):
+            # storage url too busy, change it
+            self._invalidate_upload_url(upload_url_info)
+            raise
+
+        except Exception as exc:
+            if is_temp_network_error(exc) or isinstance(exc, ssl.SSLError):
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py]: Hm.. I honestly didn't think about that, I tried to address the exceptions which Backblaze lists as cases when we should get another upload URL for a different storage pod: [https://www.backblaze.com/b2/docs/uploading.html](https://www.backblaze.com/b2/docs/uploading.html)

**Comment by kobuki** [comment]
Any chance this gets merged in the near future?

**Comment by powerpaul17** [comment]
> Any chance this gets merged in the near future?

Unfortunately I didn't have much time last month but I was working on it yesterday. I will still need some more time though..

File: src/s3ql/backends/b2/b2_backend.py [hunk 15]
```diff
@@ -0,0 +1,807 @@
...
+        self.options = options.backend_options
+
+        self.b2_application_key_id = options.backend_login
+        self.b2_application_key = options.backend_password
+
+        self.tcp_timeout = self.options.get('tcp-timeout', 20)
+
+        self.account_id = self.options.get('account-id', None)
+
+        self.disable_versions = self.options.get('disable-versions', False)
```
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py:65]: The problem is that it is possible to set more than one versioning rule for a bucket. I think it would be best to leave it as it is. In my observations it has not happened that s3ql created a new version of the same block, and users should be able to set this correctly.

[the remark this answers is not in the document]
Comment (d--j) [reply src/s3ql/backends/b2/b2_backend.py:65]: ok

**Comment by szepeviktor** [comment]
Backblaze announced its S3-compatible API
https://www.backblaze.com/blog/backblaze-b2-s3-compatible-api/

**Comment by powerpaul17** [comment]
Ok, well, that's a very nice announcement.. and I just thought about getting the rest of the comments worked out..

**Comment by kobuki** [comment]
It might still worth addressing the remaining points. I get the feeling the S3 interface is an afterthought and might not be as performant or effective as the original API. And it's also in beta for now, and who knows when it gets released as GA/final.

**Comment by szepeviktor** [comment]
> and might not be as performant or effective as the original API

It isn't for sure.

File: src/s3ql/backends/b2/b2_backend.py [hunk 16]
```diff
@@ -0,0 +1,807 @@
...
+            try:
+                self._delete_file_id(file_id['fileName'], file_id['fileId'])
+            except B2Error as exc:
+                if exc.code == 'file_not_present':
+                    # Server may have deleted the object even though we did not
+                    # receive the response.
+                    if force or is_retry:
+                        pass
+                    else:
+                        raise exc
```
[the remark this answers is not in the document]
Comment (d--j) [reply src/s3ql/backends/b2/b2_backend.py:486]: `del file_ids[:i]` while you are iterating over `file_ids` does not work as expected. So in your current implementation https://github.com/s3ql/s3ql/blob/e5af517e66c52cd1a838cc9aa9d875f0c3cd2743/src/s3ql/backends/b2/b2_backend.py#L479-L486 the `if force or is_retry` case kinda breaks the iteration (and the else case is missing the `del`).

I would suggest rewriting it like this:

```python
    @retry
    def _delete_file_ids(self, file_ids, force=False, is_retry=False):
        for (i, file_id) in enumerate(file_ids):
            try:
                self._delete_file_id(file_id['fileName'], file_id['fileId'], force or is_retry)
            except:
                del file_ids[:i]
                raise

        del file_ids[:]

    def _delete_file_id(self, file_name, file_id, force=False):
        request_dict = {
            'fileName': file_name,
            'fileId': file_id
        }
        try:
            self._do_api_call('b2_delete_file_version', request_dict)
        except B2Error as exc:
            if exc.code == 'file_not_present' and force:
                return
            raise exc
        return

```

Here the force handling is moved in the `_delete_file_id` method so that deleting a non existing file version does not raise an error at all. Also after a retry the force parameter of `_delete_file_id` is always `True` since – as you mentioned in the comments – the server response might have been lost.

**Comment by d--j** [comment]
@powerpaul17 looks to me that you very very near the finishing line! Have a look at my last comment here: https://github.com/s3ql/s3ql/pull/116/files/b5c2a547fc7df5c6fe7193ffd5d6032dc677a6b0..e5af517e66c52cd1a838cc9aa9d875f0c3cd2743#r426691485
After fixing that I think that this pull request is ready to be merged.

File: src/s3ql/backends/b2/b2_backend.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (powerpaul17) [reply src/s3ql/backends/b2/b2_backend.py:486]: Yeah, I tried to fix it yesterday and it seemed to work (at least with the test I wrote) but this is of course much better, thanks for the correction.

**Comment by d--j** [comment]
@powerpaul17 could you rebase on top of master, again? Travis needed to be fixed…

**Comment by powerpaul17** [comment]
> @powerpaul17 could you rebase on top of master, again? Travis needed to be fixed…

No problem. And done. :wink:

**Comment by szepeviktor** [comment]
All green 🍏

**Comment by d--j** [comment]
@Nikratio if you find some time: this pull request is ready to be merged from my point of view. Please have a look and merge it, if you concur.

**Comment by Nikratio** [comment]
I'll trust your judgement :-).
